src/Models/HydrostaticFreeSurfaceModels/pcg_implicit_free_surface_solver.jl

```diff
@@ -42,9 +42,14 @@ for a fluid with variable depth `H`, horizontal areas `Az`, barotropic volume fl
 step `Δt`, gravitational acceleration `g`, and free surface at time-step `n` `ηⁿ`.
 """
 function PCGImplicitFreeSurfaceSolver(grid::AbstractGrid, settings, gravitational_acceleration=nothing)
+    Hx, Hy, _ = halo_size(grid)
+
+    Hx = Hx ≤ 3 ? 3 : Hx
+    Hy = Hy ≤ 3 ? 3 : Hy
+
     # Initialize vertically integrated lateral face areas
-    ∫ᶻ_Axᶠᶜᶜ = Field((Face, Center, Nothing), with_halo((3, 3, 1), grid))
-    ∫ᶻ_Ayᶜᶠᶜ = Field((Center, Face, Nothing), with_halo((3, 3, 1), grid))
+    ∫ᶻ_Axᶠᶜᶜ = Field((Face, Center, Nothing), with_halo((Hx, Hy, 1), grid))
+    ∫ᶻ_Ayᶜᶠᶜ = Field((Center, Face, Nothing), with_halo((Hx, Hy, 1), grid))
```

Comment:
I would define 
```
adapted_grid = adapt_to_preconditioner(grid, settings.preconditioner)
∫ᶻ_Axᶠᶜᶜ = Field((Face, Center, Nothing), adapted_grid)
∫ᶻ_Ayᶜᶠᶜ = Field((Center, Face, Nothing), adapted_grid)
```

and 
```
adapt_to_preconditioner(grid, preconditioner) = grid
function adapt_to_preconditioner(grid, ::DiagonallyDominantPreconditioner)
    Hx, Hy, _ = halo_size(grid)
    Hx > 2 && Hy > 2 && return grid
    
    Ax = Hx > 2 ? Hx : 3
    Ay = Hy > 2 ? Hy : 3
    
    return with_halo((Ax, Ay, 1), grid)
end
```

Reply:
that seems sensible

Reply:
Though doesn't `inflate_halo_size` already do this? Why do we need a separate?

Something like

```julia
import Oceananigans.Grids: required_halo_size

required_halo_size(::DiagonallyDominantPreconditioner) = (3, 3, 1)

# later
new_halo = inflate_halo_size(halo_size(grid)..., grid, preconditioner)

if any(new_halo .> halo_size(grid))
    grid = with_halo(new_halo, grid)
end
```

might be better. Or, we should refactor the utilities in `Grids` so they can be used here rather than writing new stuff from scratch.

Reply:
Why does it need 1 halo in z? 0 isn't ok?

Reply:
I guess 0 should be fine

Reply:
OK, I don't understand where to make these changes and at the moment the PGC doesn't work so we need to fix it. Could you help me?

I think what I can't figure out is when and if `DiagonallyDominantPreconditioner` is chosen, then which method constructs the PCG?